### PR TITLE
Implement `approve` method

### PR DESCRIPTION
**Pull request by YoonhoKim0527**

### CODE REVIEW

File: repository/src/lib.rs [hunk 1]
```diff
@@ -439,10 +439,67 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+        // Check if the agenda is rebased on top of the `finalized` branch.
+        if self
+            .raw
+            .find_merge_base(last_header_commit, *agenda_commit_hash)
+            .await?
+            != last_header_commit
+        {
+            return Err(anyhow!(
+                "agenda:{} should be rebased on {}",
+                agenda_commit_hash,
```
Comment (TomTaehoonKim) [inline repository/src/lib.rs:456]:
```suggestion
                "branch {} should be rebased on {}",
                agenda_branch_name,
```
keep the error message same with other methods please

File: repository/src/lib.rs [hunk 2]
```diff
@@ -439,10 +439,67 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+        let agenda_commit = self.read_commit(*agenda_commit_hash).await?;
+        let agenda = match agenda_commit {
+            Commit::Agenda(agenda) => agenda,
+            _ => panic!("given commit hash is not agenda commit hash"),
+        };
+        for (_, typedsignature) in proof.clone() {
+            typedsignature.verify(&agenda)?;
+        }
+
+        // Delete past a-#(agenda) branch and create new a-#(agendaproof) branch
```
Comment (TomTaehoonKim) [inline repository/src/lib.rs:471]:
```suggestion
        // Delete past `a-(trimmed agenda hash)` branch and create new `a-(trimmed agenda proof hash)` branch
```

File: repository/src/lib.rs [hunk 3]
```diff
@@ -439,10 +439,67 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+        self.raw
+            .delete_branch(format!(
+                "a-{:?}",
+                agenda_commit_hash
+                    .to_string()
+                    .truncate(TAG_NAME_HASH_DIGITS)
```
Comment (TomTaehoonKim) [inline repository/src/lib.rs:477]:
```suggestion
                agenda_commit
                    .to_hash256()
                    .to_string()
                    .truncate(BRANCH_NAME_HASH_DIGITS)
```
wrong implementation
also add BRANCH_NAME_HASH_DIGITS please

File: repository/src/lib.rs [hunk 4]
```diff
@@ -439,10 +439,67 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+        }
+
+        // Verify agenda with agenda proof
+        let agenda_commit = self.read_commit(*agenda_commit_hash).await?;
+        let agenda = match agenda_commit {
+            Commit::Agenda(agenda) => agenda,
+            _ => panic!("given commit hash is not agenda commit hash"),
+        };
+        for (_, typedsignature) in proof.clone() {
+            typedsignature.verify(&agenda)?;
```
Comment (TomTaehoonKim) [inline repository/src/lib.rs:468]:
```suggestion
        for (_, typed_signature) in proof.clone() {
            typed_signature.verify(&agenda)?;
```
snake case please

File: repository/src/lib.rs [hunk 5]
```diff
@@ -439,10 +439,67 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+        // Create agenda proof commit
+        let agenda_commit = self.read_commit(*agenda_commit_hash).await?;
+        let agenda_proof = AgendaProof {
+            height: last_header.height + 1,
+            agenda_hash: agenda_commit.to_hash256(),
+            proof: proof.into_iter().map(|(_, sig)| sig).collect(),
+        };
+        let agenda_proof_commit = Commit::AgendaProof(agenda_proof.clone());
+        let agenda_proof_semantic_commit = format::to_semantic_commit(&agenda_proof_commit);
+        let agenda_proof_commit_hash = serde_json::from_str(&agenda_proof_semantic_commit.body)?;
```
Comment (TomTaehoonKim) [inline repository/src/lib.rs:489]: delete this wrong implementation

Comment (TomTaehoonKim) [reply repository/src/lib.rs:489]: use create_semantic_commit method and return the value returned from it
its type is CommitHash

File: repository/src/lib.rs [hunk 6]
```diff
@@ -439,10 +439,67 @@ impl<T: RawRepository> DistributedRepository<T> {
...
         &mut self,
-        _agenda_commit_hash: &CommitHash,
-        _proof: Vec<(PublicKey, TypedSignature<Agenda>)>,
+        agenda_commit_hash: &CommitHash,
+        proof: Vec<(PublicKey, TypedSignature<Agenda>)>,
     ) -> Result<CommitHash, Error> {
-        unimplemented!()
+        let last_header = self.get_last_finalized_block_header().await?;
+        let last_header_commit = self.raw.locate_branch(FINALIZED_BRANCH_NAME.into()).await?;
+        // Check if the agenda is rebased on top of the `finalized` branch.
```
[the remark this answers is not in the document]
Comment (YoonhoKim0527) [reply repository/src/lib.rs:447]: fixed.

File: repository/src/lib.rs (hunk 3 above)
Comment (YoonhoKim0527) [reply repository/src/lib.rs:477]: fixed.

File: repository/src/lib.rs (hunk 4 above)
Comment (YoonhoKim0527) [reply repository/src/lib.rs:468]: fixed.

File: repository/src/lib.rs [hunk 7]
```diff
@@ -439,10 +439,67 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+                "a-{:?}",
+                agenda_commit_hash
+                    .to_string()
+                    .truncate(TAG_NAME_HASH_DIGITS)
+            ))
+            .await?;
+        // Create agenda proof commit
+        let agenda_commit = self.read_commit(*agenda_commit_hash).await?;
+        let agenda_proof = AgendaProof {
+            height: last_header.height + 1,
```
[the remark this answers is not in the document]
Comment (YoonhoKim0527) [reply repository/src/lib.rs:483]: fixed.

File: repository/src/lib.rs [hunk 8]
```diff
@@ -439,10 +439,67 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+        self.raw
+            .delete_branch(format!(
+                "a-{:?}",
+                agenda_commit_hash
+                    .to_string()
+                    .truncate(TAG_NAME_HASH_DIGITS)
+            ))
+            .await?;
+        // Create agenda proof commit
+        let agenda_commit = self.read_commit(*agenda_commit_hash).await?;
```
[the remark this answers is not in the document]
Comment (YoonhoKim0527) [reply repository/src/lib.rs:481]: fixed.

File: repository/src/lib.rs (hunk 5 above)
Comment (YoonhoKim0527) [reply repository/src/lib.rs:489]: fixed.

File: repository/src/lib.rs (hunk 2 above)
Comment (YoonhoKim0527) [reply repository/src/lib.rs]: fixed.

File: repository/src/lib.rs (hunk 1 above)
Comment (YoonhoKim0527) [reply repository/src/lib.rs]: fixed.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: repository/src/lib.rs [hunk 9]
```diff
@@ -267,10 +268,71 @@ impl<T: RawRepository> DistributedRepository<T> {
...
-        _agenda_commit_hash: &CommitHash,
-        _proof: Vec<TypedSignature<Agenda>>,
+        agenda_commit_hash: &CommitHash,
+        proof: Vec<TypedSignature<Agenda>>,
     ) -> Result<CommitHash, Error> {
-        unimplemented!()
+        // Check if the agenda branch is rebased on top of the `finalized` branch.
+        let last_header_commit = self.raw.locate_branch(FINALIZED_BRANCH_NAME.into()).await?;
+        let agenda_branch_name = self.raw.get_branches(*agenda_commit_hash).await?;
+        let agenda_branch_name = agenda_branch_name.get(0).unwrap();
```
Comment (junha1) [inline repository/src/lib.rs:277]: no `unwrap` here. This is just an argument error

File: repository/src/lib.rs [hunk 10]
```diff
@@ -267,10 +268,71 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+                agenda_branch_name,
+                FINALIZED_BRANCH_NAME
+            ));
+        }
+
+        // Verify agenda with agenda proof
+        let agenda_commit = self.read_commit(*agenda_commit_hash).await?;
+        let agenda = match agenda_commit.clone() {
+            Commit::Agenda(agenda) => agenda,
+            _ => panic!("given commit hash is not agenda commit hash"),
```
Comment (junha1) [inline repository/src/lib.rs:295]: just handle the error

File: repository/src/lib.rs [hunk 11]
```diff
@@ -267,10 +268,71 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+        };
+        for typed_signature in proof.clone() {
+            typed_signature.verify(&agenda)?;
+        }
```
Comment (junha1) [inline repository/src/lib.rs:299]: This is not enough. You must use CSV, starting from the finalized block.

File: repository/src/lib.rs [hunk 12]
```diff
@@ -267,10 +268,71 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+
+        // Delete past `a-(trimmed agenda hash)` branch and create new `a-(trimmed agenda proof hash)` branch
+        self.raw
+            .delete_branch(format!(
```
Comment (junha1) [inline repository/src/lib.rs:303]: You already have `agenda_branch_name`

File: repository/src/lib.rs (hunk 9 above)
Comment (YoonhoKim0527) [reply repository/src/lib.rs]: fixed.

File: repository/src/lib.rs (hunk 11 above)
Comment (YoonhoKim0527) [reply repository/src/lib.rs:336]: fixed. I added csv before this.

File: repository/src/lib.rs (hunk 12 above)
Comment (YoonhoKim0527) [reply repository/src/lib.rs]: fixed.

File: repository/src/lib.rs [hunk 13]
```diff
@@ -269,10 +270,79 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+        let mut verifier = CommitSequenceVerifier::new(finalized_header.clone(), reserved_state)
+            .map_err(|e| anyhow!("failed to create a verifier: {}", e))?;
+        for (commit, hash) in commits.iter() {
+            verifier
+                .apply_commit(commit)
+                .map_err(|e| anyhow!("verification error on commit {}: {}", hash, e))?;
+        }
+
+        // Verify agenda with agenda proof
+        let agenda_commit = self.read_commit(*agenda_commit_hash).await?;
```
[the remark this answers is not in the document]
Comment (YoonhoKim0527) [reply repository/src/lib.rs]: fixed.

File: repository/src/lib.rs [hunk 14]
```diff
@@ -269,10 +270,75 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+                .apply_commit(commit)
+                .map_err(|e| anyhow!("verification error on commit {}: {}", hash, e))?;
+        }
+        // Verify agenda with agenda proof
+        let agenda_commit = commits.iter().map(|(commit, _)| commit).last().unwrap();
+        let agenda = match agenda_commit.clone() {
+            Commit::Agenda(agenda) => Some(agenda),
+            _ => None,
+        };
+        let agenda = agenda.ok_or_else(|| anyhow::anyhow!("not an agenda commit"))?;
```
Comment (junha1) [inline repository/src/lib.rs:313]:
```rust
let agenda = match agenda_commit { Commit::Agenda{agenda} => agenda, _ => return anyhow ... };
```

Comment (YoonhoKim0527) [reply repository/src/lib.rs:313]: fixed.

**Review by junha1 — APPROVED** [review]

**Pull request merged**